app/services/katello/registration_manager.rb

```diff
@@ -0,0 +1,197 @@
+          end
+        end
+        host.host_collection_ids = host_collection_ids
+      end
+
+      def lookup_content_view_environment(activation_keys)
+        activation_key = activation_keys.reverse.detect do |act_key|
+          act_key.environment && act_key.content_view
+        end
+        if activation_key
+          ::Katello::ContentViewEnvironment.where(:content_view_id => activation_key.content_view, :environment_id => activation_key.environment).first
+        else
+          fail _('At least one activation key must have a lifecycle environment and content view assigned to it')
+        end
+      end
+
+      def candlepin_consumer_destroy(host_uuid)
+        ::Katello::Resources::Candlepin::Consumer.destroy(host_uuid)
+      rescue RestClient::Gone
+        Rails.logger.error(_("Consumer %s has already been removed") % host_uuid)
+      end
+
+      def pulp_consumer_destory(host_uuid)
+        ::Katello.pulp_server.extensions.consumer.delete(host_uuid)
+      rescue RestClient::ResourceNotFound
+        Rails.logger.error(_("Pulp Consumer %s has already been removed") % host_uuid)
+      end
+
+      def set_content_and_subscription_uuids(host, uuid)
+        host.content_facet.uuid = uuid
+        host.subscription_facet.uuid = uuid
+        host.subscription_facet.user = User.current unless User.current.nil? || User.current.hidden?
+        host.content_facet.save!
+        host.subscription_facet.save!
+      end
+
+      def populate_content_facet(host, content_view_environment)
+        content_facet = host.content_facet || ::Katello::Host::ContentFacet.new(:host => host)
+        content_facet.content_view = content_view_environment.content_view
+        content_facet.lifecycle_environment = content_view_environment.environment
+        content_facet.save!
+        content_facet
+      end
+
+      def populate_subscription_facet(host, activation_keys, consumer_params)
+        subscription_facet = host.subscription_facet || ::Katello::Host::SubscriptionFacet.new(:host => host)
+        subscription_facet.last_checkin = Time.now
+        subscription_facet.update_from_consumer_attributes(consumer_params.except(:installedProducts, :guestIds, :facts))
+        subscription_facet.save!
+        subscription_facet.activation_keys = activation_keys
+        subscription_facet
+      end
+
+      def remove_host_artifacts(host)
+        if host.content_facet
+          host.content_facet.bound_repositories = []
+          host.content_facet.applicable_errata = []
+          host.content_facet.uuid = nil
+          host.content_facet.save!
+        end
+        destroy_host_artifacts(host)
+      end
+
+      def destroy_host_artifacts(host)
+        host.get_status(::Katello::ErrataStatus).destroy
+        host.get_status(::Katello::SubscriptionStatus).destroy
+        host.get_status(::Katello::TraceStatus).destroy
+        host.installed_packages.destroy_all
+      end
```

Comment:
I think these two method names are confusing, any way to make them more clear?   Maybe "destroy_host_artifacts" and "destroy_host_and_facet_artifacts" ?  Or maybe something like:

```
def destroy_host_artifacts(clear_content_facet => false)
end
```

and then you just flip that based on where its called from?